Sync on common config elements during startup

dkales opened the pull request:
This PR computes a hash of the common config elements and syncs on it during startup to detect basic config mismatches.

TODOs:
- [ ] Agree on all parts of config which are part of sync
- [x] ATM just a basic hash is sent around, should we just send around a common config struct to print better errors?

carlomazzaferro commented on iris-mpc-common/src/config/mod.rs:
```diff
@@ -211,6 +212,121 @@ pub struct Config {
+            environment,
+            party_id: _,           // party id is different for each server
+            requests_queue_url: _, // requests queue url is different for each server
+            results_topic_arn,     // TODO: do we want the results topic ARN in the hash?
```
Would say yes

carlomazzaferro commented on iris-mpc-common/src/config/mod.rs:
```diff
@@ -211,6 +212,121 @@ pub struct Config {
+            aws: _,          // aws is different for each server
+            processing_timeout_secs,
+            startup_sync_timeout_secs,
+            public_key_base_url: _, // TODO: include in common config hash?
```
also yes

carlomazzaferro commented on iris-mpc-common/src/config/mod.rs:
```diff
@@ -211,6 +212,121 @@ pub struct Config {
+            processing_timeout_secs,
+            startup_sync_timeout_secs,
+            public_key_base_url: _, // TODO: include in common config hash?
+            shares_bucket_name: _,  // TODO: should this be the same for all servers?
```
yes, same for all

carlomazzaferro commented on iris-mpc-common/src/config/mod.rs:
```diff
@@ -211,6 +212,121 @@ pub struct Config {
+            return_partial_results,
+            disable_persistence,
+            enable_debug_timing: _,
+            node_hostnames, // TODO: is this the same for all servers?
```
yes, all the same

carlomazzaferro replied:
Except... When running locally. Let's leave it out for now

carlomazzaferro replied:
See: https://github.com/worldcoin/iris-mpc/blob/36b5b4732ab93b767de254eb2b5b2858ca1f0035/.test.hawk0.env#L3

We'd have to change this, otherwise locally this will fail. Let's leave it out for now.

naure replied:
Yes, this 0.0.0.0 is not necessary.

carlomazzaferro commented on iris-mpc-common/src/config/mod.rs:
```diff
@@ -211,6 +212,121 @@ pub struct Config {
+            enable_reauth,
+            hawk_request_parallelism,
+            hawk_connection_parallelism,
+            hawk_server_healthcheck_port: _, // different for each server
```
Only locally, but I get your point

dkales commented:
Cleaned this up a bit, and removed some things from the common sync that might have a good reason to be different (ports & urls due to proxies/LB/etc).

BTW the CPU integration tests cut of the logs for the docker-compose up step, so the reason for the error (the new startup sync failed) was not apparent in the previous runs. IDK if this can be fixed or if we would need to not use the action.

naure commented:
It seems easier and more useful to send the whole config indeed.
